src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -227,6 +238,813 @@ class OMPSerialDirective(OMPRegionDirective, metaclass=abc.ABCMeta):
+            member = node1
+            member1 = node2
+
+            # We can't just do == on the Member child, as that
+            # will recurse and check the array indices for any
+            # ArrayMixin children
+
+            # Check the signature of both StructureReference
+            # to see if they are accessing the same data
+            ref0_sig = node1.get_signature_and_indices()[0]
+            ref1_sig = node2.get_signature_and_indices()[0]
+            if ref0_sig != ref1_sig:
+                return True
+
+        # If we have (exactly) Reference objects we filter out
+        # non-matching ones with the symbol check, and matching ones
+        # are always valid since they are simple accesses.
+        if type(node1) is Reference:
+            return True
+
+        # All remaining objects are some sort of Array access
+        array1 = None
+        array2 = None
+        if isinstance(node1, ArrayReference):
+            array1 = node1
+            array2 = node2
+        else:
+            array1 = node1.walk(ArrayMixin)[0]
+            array2 = node2.walk(ArrayMixin)[0]
+        valid = True
+        for i, index in enumerate(array1.indices):
+            if (isinstance(index, Literal) or
+                    isinstance(array2.indices[i], Literal)):
+                valid = valid and self._valid_dependence_literals(
+                            index, array2.indices[i])
+            elif (isinstance(index, Range) or
+                  isinstance(array2.indices[i], Range)):
+                valid = valid and self._valid_dependence_ranges(
+                            array1, array2, i)
+            else:
+                # The only remaining option is that the indices are
+                # References or BinaryOperations
+                valid = valid and self._valid_dependence_ref_binop(
+                            index, array2.indices[i], task1, task2)
+
+        return valid
+
+    def _validate_task_dependencies(self):
+        '''
+        Validates all task dependencies in this OMPSerialDirective region are
+        valid within the restraints of OpenMP & PSyclone. This is done through
+        a variety of helper functions, and checks each pair of tasks' inout,
+        outin and outout combinations.
+
+        Any task dependencies that are detected and will not be handled by
+        OpenMP's depend clause will be handled through the addition of
+        OMPTaskwaitDirective nodes.
+
+        :raises NotImplementedError: If this region contains both an \
+                                 OMPTaskDirective and an OMPTaskloopDirective.
+        '''
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.nodes.omp_task_directive import OMPTaskDirective
+        tasks = self.walk(OMPTaskDirective)
+        # For now we disallow Tasks and Taskloop directives in the same Serial
+        # Region
+        if len(tasks) > 0 and len(self.walk(OMPTaskloopDirective)) > 0:
```

Comment:
Probably, is that how that syntax is optimised? Does it make the walk call stop early? I've made the change anyway.

Reply:
We probably need to change our implementation of `walk` so that it does `yield`. Not something for this PR (but I agree that using `any` does at least allow us to benefit in future).